packages/server/src/api/rest/index.ts

```
@@ -220,29 +236,60 @@ class RequestHandler extends APIHandlerBase {
+        );
+        this.urlPatternMap = this.buildUrlPatternMap(segmentCharset);
     }
 
-    buildUrlPatterns(idDivider: string, urlSegmentNameCharset: string) {
+    private buildUrlPatternMap(urlSegmentNameCharset: string): Record<UrlPatterns, UrlPattern> {
         const options = { segmentValueCharset: urlSegmentNameCharset };
+
+        const buildPath = (segments: string[]) => {
+            return (this.prefix ?? '') + '/' + segments.join('/');
+        };
+
         return {
-            // collection operations
-            collection: new UrlPattern('/:type', options),
-            // single resource operations
-            single: new UrlPattern('/:type/:id', options),
-            // related entity fetching
-            fetchRelationship: new UrlPattern('/:type/:id/:relationship', options),
-            // relationship operations
-            relationship: new UrlPattern('/:type/:id/relationships/:relationship', options),
+            [UrlPatterns.SINGLE]: new UrlPattern(buildPath([':type', ':id']), options),
+            [UrlPatterns.FETCH_RELATIONSHIP]: new UrlPattern(buildPath([':type', ':id', ':relationship']), options),
+            [UrlPatterns.RELATIONSHIP]: new UrlPattern(
+                buildPath([':type', ':id', 'relationships', ':relationship']),
+                options
+            ),
+            [UrlPatterns.COLLECTION]: new UrlPattern(buildPath([':type']), options),
         };
     }
 
+    private reverseModelNameMap(type: string): string {
+        return this.reverseModelNameMapping[type] ?? type;
+    }
+
+    private matchUrlPattern(path: string, routeType: UrlPatterns): Match {
+        const pattern = this.urlPatternMap[routeType];
+        if (!pattern) {
+            throw new InvalidValueError(`Unknown route type: ${routeType}`);
+        }
+
+        const match = pattern.match(path);
+        if (match) {
+            match.type = this.modelNameMapping[match.type] ?? match.type;
```

When a mapping is provided,  say "users" => "user", should access via unmapped path name like "POST /api/user" fail? I tend to think so - configuring a mapping requires you to only use the mapped name.

What do you think? Currently I believe both are valid.

---

Indeed. It should fail. We only had the case in mind where only some of the models are mapped but not all.

So should we rather throw an exception if a model is not mapped if there are any mappings?

---

I now did some testing. At least with Nest.js it already returns a 404, since there is no `/user` route defined. Now if someone would then also define a `/user` route, indeed without change it would then work but at that point that seems to be intended by the user. So imho there is no need to handle this case inside the `RequestHandler`.

---

My understanding is the mapping (in the latest code) is inverted.

I tested the latest code with this mapping:

```ts
{
    user: 'myUser',
    post: 'myPost',
}
```

And this test is failing because of returning 200:
```ts
            await expect(
                handler({
                    method: 'post',
                    path: '/user',
                    query: {},
                    requestBody: { data: { type: 'user', attributes: { name: 'User1' } } },
                    prisma,
                })
            ).resolves.toMatchObject({
                status: 400,
            });
```